### PR TITLE
[RFC] SQL Type Annotations

### PR DESCRIPTION
**Pull request by nvb**
[ci skip]

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/6895)

### CODE REVIEW

**Comment by bdarnell** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/type_annotations.md, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIdns2g4jQRwlCSXbQ5:-KIdns2g4jQRwlCSXbQ6:713607753) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L222)):*
> ```Markdown
> portability of SQL statements and create a higher learning curve for
> moving to Cockroach. However, we concluded that this was ok, because 
> these annotations are fully op-in, and are not a requirement of the 
> ```

s/op/opt/

---

*[docs/RFCS/type_annotations.md, line 240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIdr2ngUcEm_bv-owR-:-KIdr2ngUcEm_bv-owR0:-1491258275) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L240)):*
> ```Markdown
> `E : T`
> - [+] has parallels to Postgres' cast extension `E :: T`
> - [-] may look too similar to Postgres' cast extension
> ```

`:` is also used for placeholders in oracle.

---

*[docs/RFCS/type_annotations.md, line 250 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIdnuVIPyEcZr5ioXP2:-KIdnuVIPyEcZr5ioXP3:256011267) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L250)):*
> ```Markdown
> - [+] a mix between the previous two proposals
> - [+] unconventional (wont be mistaken)
> - [-] unconventional (stands out)
> ```

May also be confused with unary `!`, which means logical negation in many languages including mysql, and factorial in postgresql. Postgres even has a unary `!!` (which also means factorial - single `!` is a postfix operator; `!!` is prefix). 

---

*[docs/RFCS/type_annotations.md, line 261 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIdpQ_PQZkUsZDH1XMb:-KIdpQ_PQZkUsZDH1XMc:1528511963) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L261)):*
> ```Markdown
> desired that we adopt a more standard keyword style syntax.
> 
> For instance: `ANNOTATE(E AS T)` or `(E ANNOTATE AS T)`.
> ```

It's also worth considering something that looks like a plain function: `ANNOTATE(E, T)`. This is easier to plumb through interfaces like SQLAlchemy's query builder (which has a special case for `CAST`, but it's difficult to add other things with similar treatment. For that matter, it's also difficult to use non-standard operators). There is some precedent for using a type as an ordinary function parameter in mysql's `CONVERT(E, T)` function, although I'm not sure how happy our grammar would be with that.

My preference is for the plain function syntax because that's going to be easiest to pass through client layers. I'd rather not introduce multiple syntaxes for type annotations unless they turn out to be common enough to warrant a short operator form.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895)*

File: docs/RFCS/type_annotations.md [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+In order to clarify the typing rules during the design of the Summer
+type system and to exercise the proposed system, we found it was useful
+to "force" certain expressions to be of a certain type.
+
+Unfortunately, the SQL cast expression (`CAST(... AS ...)` or
+`...::...`) is not appropriate for this, because although it
+guarantees a type to the surrounding expression it does not constrain
+its argument. For example `sign(1.2)::int` does not disambiguate which
+overload of `sign` to use.
```
Comment (tschottdorf) [inline docs/RFCS/type_annotations.md]: is `sign($1::int)` valid, and would it do what you want here?

**Comment by RaduBerinde** [comment]
This sounds great to me. I especially like that it can be used to force a placeholder to have a certain type - with the power of that and casts, one should always be able to get around any deficiency in the typing system.

I like the "x : type" syntax, and I agree that just like :: has a more explicit `CAST(.. AS ..)` syntax this should also have an explicit alias, e.g.  `SETTYPE(... , ...)`

<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> #### \[RFC\] SQL Type Annotations
> 
> [ci skip]
> 
> 
> 
</details>


---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/type_annotations.md, line 114 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIeyv616RdKgH_k_nbn:-KIeyv62e5C77mw10nGx:1571388391) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L114)):*
> ```Markdown
>         -> no preference for the return type of `1.2`
>         -> `1.2` defaults to a DECIMAL
>         -> overload resolution choses the DECIMAL implementation
> ```

chooses (and two more below)

---

*[docs/RFCS/type_annotations.md, line 261 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIdpQ_PQZkUsZDH1XMb:-KIf0QDJUndlw_pj4uHs:-737692153) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L261)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It's also worth considering something that looks like a plain function: `ANNOTATE(E, T)`. This is easier to plumb through interfaces like SQLAlchemy's query builder (which has a special case for `CAST`, but it's difficult to add other things with similar treatment. For that matter, it's also difficult to use non-standard operators). There is some precedent for using a type as an ordinary function parameter in mysql's `CONVERT(E, T)` function, although I'm not sure how happy our grammar would be with that.
> 
> My preference is for the plain function syntax because that's going to be easiest to pass through client layers. I'd rather not introduce multiple syntaxes for type annotations unless they turn out to be common enough to warrant a short operator form.
</details>
I like the function-style syntax.  But `ANNOTATE` is pretty obtuse, how about `SETTYPE(E, T)` or `SET_TYPE()` or `WITH_TYPE()`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-:-KIezUbZwrbN_1oTqNN5:-1160256023)*

**Comment by knz** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/type_annotations.md, line 27 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIcdBYqoc1vAnA_aUxV-r1-27:-KJ6TPEVCdZOD7L9cQPz:1813846216) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L27)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> is `sign($1::int)` valid, and would it do what you want here?
</details>
Yes in that particular case there is a way to solve the problem. Perhaps the more general statement would be "in the case where there are multiple overloads of a function all with the same return type, a cast on the return type has no effect on overload resolution". Or something to that effect.

---

*[docs/RFCS/type_annotations.md, line 156 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KJ6WBJ1egtDwNMJYEpV:-KJ6WBJ30dbnczkHVEry:-555367501) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L156)):*
> ```Markdown
> 
> The initial Summer RFC defined a set of rules for determining the type of
> placeholders based on type casts and type annotations. These rules were:
> ```

Perhaps put this in present tense (or say "the proposed rules are")

---

*[docs/RFCS/type_annotations.md, line 261 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIdpQ_PQZkUsZDH1XMb:-KJ6WYQBdm_BoPZePZti:-1028381781) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L261)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> I like the function-style syntax.  But `ANNOTATE` is pretty obtuse, how about `SETTYPE(E, T)` or `SET_TYPE()` or `WITH_TYPE()`
</details>
+1 on a function-like syntax. I don't like "set type" much though. "with_type" is ok, also consider "typed_as(E, T)", perhaps even "typed(E, T)" .

Btw I think we should have both a function like syntax and also an operator for conciseness in docs. I still like the single ":" best, visual ambiguity with "::" notwithstanding (and actually the closeness with "::" might be an advantage).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895)*

**Comment by nvanbenschoten** [comment]
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/type_annotations.md, line 114 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIeyv616RdKgH_k_nbn:-KJ7UWr8yYXSXHHRakpy:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L114)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> chooses (and two more below)
</details>
Done.

---

*[docs/RFCS/type_annotations.md, line 156 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KJ6WBJ1egtDwNMJYEpV:-KJ7UgfCfZ6QNj0Kdk1-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L156)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> Perhaps put this in present tense (or say "the proposed rules are")
</details>
Done.

---

*[docs/RFCS/type_annotations.md, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIdns2g4jQRwlCSXbQ5:-KJ7UiYy11pQcXnvWPmJ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L222)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/op/opt/
</details>
Done.

---

*[docs/RFCS/type_annotations.md, line 240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIdr2ngUcEm_bv-owR-:-KJ7WOGesesaIldJYwZL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L240)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> `:` is also used for placeholders in oracle.
</details>
Done.

---

*[docs/RFCS/type_annotations.md, line 250 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIdnuVIPyEcZr5ioXP2:-KJ7WcshC5yODTr6NSS5:682180434) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L250)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> May also be confused with unary `!`, which means logical negation in many languages including mysql, and factorial in postgresql. Postgres even has a unary `!!` (which also means factorial - single `!` is a postfix operator; `!!` is prefix). 
</details>
Good point, noted in the cons lists.

---

*[docs/RFCS/type_annotations.md, line 261 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIdpQ_PQZkUsZDH1XMb:-KJ7WtRVdpspG2GZIjI7:-2006956750) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L261)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary>
> +1 on a function-like syntax. I don't like "set type" much though. "with_type" is ok, also consider "typed_as(E, T)", perhaps even "typed(E, T)" .
> 
> Btw I think we should have both a function like syntax and also an operator for conciseness in docs. I still like the single ":" best, visual ambiguity with "::" notwithstanding (and actually the closeness with "::" might be an advantage).
</details>
Syntax that looks like a function is a good idea. With regards to interfaces like SQLAlchemy's query builder, is it possible to provide a type name as an argument? If not, we may need to accept strings for the `T` arg instead/as well.

I think I like `WITH_TYPE(E, T)` the best out of all of the proposed names. I also agree with @knz's proposal to have a function like syntax and an operator for conciseness. I'd rather not mix the logical concepts of what function's role is is vs. what an operator's role is, and if we only have a function, I think it might cause some confusion. If we follow this route, the name of the function should probably be somewhat consistent with what we call the operator. Because of this, either `TYPE_ANNOTATE` or `ANNOTATE_TYPE` might also be good choices for the function name.


On a side note @bdarnell, you mentioned that "it's also difficult to use non-standard operators". Does SQLAlchemy expose floor division? I would assume it needs some way to access MySQL's `DIV` operator. If so, I'll look into how we can customize our plugin to work with our new `//` operator.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895)*

**Comment by bdarnell** [comment]
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/type_annotations.md, line 261 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895#-KIdpQ_PQZkUsZDH1XMb:-KJ7bq5LjV4r1fgZd4u9:-1490731375) ([raw file](https://github.com/cockroachdb/cockroach/blob/80479e9f3536ff571b251549219e1499afad78e7/docs/RFCS/type_annotations.md#L261)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary>
> Syntax that looks like a function is a good idea. With regards to interfaces like SQLAlchemy's query builder, is it possible to provide a type name as an argument? If not, we may need to accept strings for the `T` arg instead/as well.
> 
> I think I like `WITH_TYPE(E, T)` the best out of all of the proposed names. I also agree with @knz's proposal to have a function like syntax and an operator for conciseness. I'd rather not mix the logical concepts of what function's role is is vs. what an operator's role is, and if we only have a function, I think it might cause some confusion. If we follow this route, the name of the function should probably be somewhat consistent with what we call the operator. Because of this, either `TYPE_ANNOTATE` or `ANNOTATE_TYPE` might also be good choices for the function name.
> 
> 
> On a side note @bdarnell, you mentioned that "it's also difficult to use non-standard operators". Does SQLAlchemy expose floor division? I would assume it needs some way to access MySQL's `DIV` operator. If so, I'll look into how we can customize our plugin to work with our new `//` operator.
</details>
I don't think sqlalchemy makes it easy to provide a type name as an argument, although I haven't explored this in detail. It does provide the ability to pass unquoted strings (by wrapping them in its `literal()` function) , so that's available as a last resort.

I was wrong about custom operators - there is a [method](http://docs.sqlalchemy.org/en/latest/core/sqlelement.html#sqlalchemy.sql.expression.ColumnElement.op) for this. The syntax is ugly, though (`x.op('//')(y)` for floor division), so function-style syntax is still preferable to custom operators. For floor division, since Python happens to have `//` as well, it would be possible to hook up the python operator to our `//` and mysql's `DIV`, but this would require changes to sqlalchemy proper; it's not something a plugin can do as far as I can tell.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6895)*

**Comment by nvanbenschoten** [comment]
I'm still working on a shorthand syntax for this functionality, because `:` collides with array slicing. One other idea was `Expr @: Type`. This is kind of a crazy syntax, but would be a somewhat natural fit for the name "type annotation". Any opinions?

For now though, I'm going to try to move along the first iteration of the type annotation code through without an accompanying shorthand notation. See https://github.com/cockroachdb/cockroach/pull/7083.

**Comment by nvanbenschoten** [comment]
@knz Updated the RFC to include the operator syntax implemented in #9086.

**Comment by knz** [comment]
LGTM

Any work remaining to be done before we can merge this one?

**Comment by nvanbenschoten** [comment]
I think it's ready to merge!

**Pull request merged**